### DISCUSSION TITLE
Jeremy's Idea Pad

### DISCUSSION THREAD

**Discussion by jermatic** [zauberzeug/nicegui#2920] [Ideas / Feature Requests]
- On the fly
- busy working
- hope this helps
- goal to devote company resources to providing PRs one day
-  if there is a solution to the target use case of an idea please comment it
- if my idea is a hard miss from some angle, cest la vie
- if there is already a discussion or gh issue for an idea, sweet
- if you want me to points on why an idea is worth it, i will provide as i have time but its not guaranteed
- nicegui maximalist

---------------------------------------------------------------------------------------------------------------------

### First Class Support for
- polars
- typescript

### Others
- context managers for common operations, ex disable/enable
- interface for using server other than fastapi
- more js frameworks react/svelte/etc
- more generics less anys
- ability to run python on client side in wasm

4/21/24
- shadcn styling system
https://www.shadcn-vue.com/
have a central place in code to alter the styling and more of foundational components
- vsc extension
bring the power of the IDE to nicegui
- browser extensions
- mobile native bindings (via react-native?)

4/23/24
replace run_cpu_bound with rust threading. get true multithreading without the overhead n hassle of multiprocessing.
his method is a huge headache and time sink but very necessary to ensure not blocking

side note: in general start replacing backbone python code with rust (pyo3+maturin)
no gil python is coming but not faster than offloading heavy lifting to rust

**Comment by falkoschindler** [comment]
Thanks, @jermatic!

+1 for more generics, see https://github.com/zauberzeug/nicegui/discussions/2677 👍🏻

Other suggestions like TypeScript, non-FastAPI, React/Svelte/..., client-side Python sound pretty ambitious, but _might_ be possible some day.

**Reply by jermatic** [reply]
haha, yeah I know some are very ambitious but this project does it right so worth it to me long term to get the ideas wrote down. I am doing a rewrite of my company's app with my team, so I think we can assist with the more ambitious ones along the way in some manner when we have free resources to devote to such an endeavor .. ie we will implement stuff adjacent to these items locally for our needs then pass along what we learn, maybe even have a robust enough impl to provide code.

**Reply by jermatic** [reply]
If i had to rank order the more ambitious ones .. i think typescript i will implement locally pretty soon just to make the development process easier on my devs with such a big app and all the clientside js we have to write.

**Comment by rodja** [comment]
Some of these are already tracked in separate feature requests. Is it ok if we close this one @jermatic? We favour individual request because they are simpler to track and manage... especially upvoting is not possible with such "collections".

**Reply by jermatic** [reply]
as long as closing does effect my ability to edit the list up top and do all things .. this is just my idea pad, not targeting votes or immediate support n such .. just a place to consolidate my ideas with visibility

**Reply by jermatic** [reply]
if someone wants to copy/paste one of my ideas into a single ticket for tracking they can keep the votes it gets haha

**Comment by jermatic** [comment]
update: 4/23/24

naive typescript support added locally.
using swc to transpile ts to js bundle and chokidar swc integration so swc can watch files .. then either provide js paths for components or `Path(path_to_js_file).read_text()` to use source file instead of raw js string in python.

immediate todo:
- integrate with ui.run so that watching starts when nicegui starts 
how to handle multiple nicegui apps running without watching files not used by that app?
uvicorn handles watching so i need to start there. Initial thought, script that traverses FS starting at nicegui entry file being used via looking at imports for python, make dynamic such that when a new file is added and upon import gets added to files to watch, also reverse op if import is removed. seems feasible just time to do the work for it.
